### ISSUE TITLE
[iOS Login] App crashes on login iOS11

### ISSUE THREAD

**Issue by malekelkamel** [jeduan/cordova-plugin-facebook4#640]
Hi guys!
I' am facing an issue with the Facebook login on iOS 11 . The app crashes when I select connect to Facebook using the Facebook app but it seems to be working just fine when I connect using the email address and password. Any ideas on how to fix this issue ? I'am using the latest version of the plugin .
Thnx 👍

**Comment by darkthsideous** [comment]
So did faceboook revoke this plugins access?

**Comment by peterpeterparker** [comment]
@darthsideous I can't reproduce the problem, described by OP, could you?

**Comment by darkthsideous** [comment]
@peterpeterparker Hi. Well im having an issue for iphones on IOS11 and higher. Wont allow me to retrieve any data and crashes. Is it due to ios 11 policy changes

**Comment by peterpeterparker** [comment]
@darkthsideous interesting, like I said, I don't face such problems

Could u try with the master branch of the plugin?

**Comment by darkthsideous** [comment]
@peterpeterparker definitely i'll try and let you know

**Comment by darkthsideous** [comment]
@peterpeterparker  Just ot be clear you're able to pull data from Facebook public profile on an IOS 11 device

**Comment by peterpeterparker** [comment]
@darkthsideous my app only works with a Facebook login (mandatory) I still see users successful loggin in and registering and an update of my app went thru the Apple review process last Sunday

But you never know you know

Master branch of the plugin is updated with the last Facook iOS Sdk recommend by Fb for iOS 11

**Comment by darkthsideous** [comment]
@peterpeterparker ok definitely i'll try it out. Thank you so much!!

**Comment by xavier-villelegier** [comment]
@peterpeterparker Just curious to know, is there any plan to release the `master` branch on npm soon ? 🤔I'm not really confident by using a branch as a dependency for production and not stick to a sable release (currently 1.9.1). I see that some interesting PRs about iOS 11 were merged in the last months

**Comment by peterpeterparker** [comment]
@xavier-villelegier I'm waiting on a feedback to see if we pack some major improvements for cordova-android too in the same version of if I release only the update for the iOS SDK

I definitely have an eye on it

p.s.: I'm still running my production with current v1.9.1 and didn't notice any problem yet. waiting on feedbacks about this too

note: update of the SDK will introduce a change in the login flow by adding an additional popup in it which isn't user friendly (see CHANGELOG)

**Comment by xavier-villelegier** [comment]
@peterpeterparker Thanks for the quick response, and for the details, I didn't know about the additional popup ! I'll stick to v1.9.1 for the moment then 👍

**Comment by peterpeterparker** [comment]
@xavier-villelegier as long as @darkthsideous or anyone else don't detect really major issues and as long as Apple or Facebook don't rush to force the update, I'm agree, v1.9.1 is fine

still would be cool to publish the version in npm just in case for the future you know...like I said I keep an eye on it

**Comment by monsif** [comment]
@peterpeterparker any update on the new version ? i am experience the same problems on IOS 11, i tested the same on IOS 10 and it went well.

**Comment by peterpeterparker** [comment]
@monsif could you try again with the master branch of the plugin? do you then face the error on iOS11 too or not?

till today no one confirmed me that this iOS 11 issue is effectively solved by updating the Facebook SDK for iOS (which is already updated in the master branch)

of course if you confirm me that and it's the case, I gonna then stress the release

but first I would really appreciate a feedback on this

**Comment by monsif** [comment]
@peterpeterparker i could not update the plugin from master, i've tried the following: 
npm install git://github.com/jeduan/cordova-plugin-facebook4.git --save
npm install git+https://github.com/jeduan/cordova-plugin-facebook4.git --save
npm install git+ssh://github.com/jeduan/cordova-plugin-facebook4.git --save
please advise.

**Comment by monsif** [comment]
@peterpeterparker i managed finally to install the update, i think i am doing something wrong because Facebook does not open even for permission, can you please check the following code or provide a sample code.

  ```
declare var facebookConnectPlugin;
@Component({
...
public loginFB() {
    facebookConnectPlugin.login(['public_profile', 'user_birthday', 'user_friends', 'email'], function (response) {
      let userId = response.authResponse.userID;
      let params = new Array<string>();
      facebookConnectPlugin.api("/me?fields=name,gender,birthday,email", params, function onSuccess(profile) {
        console.log("checking user ");
        this.auth.checkUser(profile.email).subscribe((result)=> {
          if (result) {
            console.log("redirecting ");
            this.redirectAfterLogin();
          } else {
            console.log("creating profile ");
            this.createProfile(profile, userId);
          }
        });
      }, function onError(error) {
        console.error("Failed: ", error);
      })
    }, function (error) {
      console.error("Failed: ", error);
    });
  }
```

**Comment by peterpeterparker** [comment]
@monsif if the problem is not related to iOS11 you should open an issue on stackoverflow. also like this it's a little bit difficult to because you didn't document the stacktrace of your error. only thing I saw is that you should use the accessToken in your api call.

**Comment by peterpeterparker** [comment]
the master branch, v1.10.0, including the update of the Facebook SDK for iOS has been now released to npm 

ps.: @xavier-villelegier I tested again today, with the last iOS 11.3 the modal "problem" is solved. only one modal gonna be presented to the user, the login flow looks even better and easier now

**Comment by peterpeterparker** [comment]
I gonna close this issue.

I wasn't able to reproduce the problem, I just tested several times, the plugin is compatible with iOS >= 11 and like I said above, the last SDK has been released

If you still face error plz provide more details about the error, demo APP repo or even stacktrace in order to be able to reproduce the problem

**Comment by fzamperin** [comment]
I'm using ionic 3 with v1.10.0 of the plugin, my app crashes and this is the stack trace of iOS debugger, tested on a iOS 11.2 simulator iphone-se

```
2018-04-14 17:54:26.911378-0300 Massa[45435:2792576] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSTaggedPointerString containsObject:]: unrecognized selector sent to instance 0xa000000000062662'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010658f12b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x00000001054abf41 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000106610024 -[NSObject(NSObject) doesNotRecognizeSelector:] + 132
	3   CoreFoundation                      0x0000000106511f78 ___forwarding___ + 1432
	4   CoreFoundation                      0x0000000106511958 _CF_forwarding_prep_0 + 120
	5   Massa                               0x0000000101cd33aa +[FBSDKInternalUtility isRegisteredCanOpenURLScheme:] + 61
	6   Massa                               0x0000000101cd32a0 +[FBSDKInternalUtility checkRegisteredCanOpenURLScheme:] + 158
	7   libdispatch.dylib                   0x000000010961b33d _dispatch_client_callout + 8
	8   libdispatch.dylib                   0x000000010961c8c8 dispatch_once_f + 297
	9   Massa                               0x0000000101cd1d8e +[FBSDKInternalUtility isFacebookAppInstalled] + 67
	10  Massa                               0x0000000101cfde4a -[FBSDKLoginManager logInParametersWithPermissions:serverConfiguration:] + 360
	11  Massa                               0x0000000101cfe29e -[FBSDKLoginManager logInWithBehavior:] + 87
	12  Massa                               0x0000000101cfe22a -[FBSDKLoginManager logInWithPermissions:handler:] + 312
	13  Massa                               0x0000000101cfcce4 -[FBSDKLoginManager logInWithReadPermissions:fromViewController:handler:] + 304
	14  Massa                               0x0000000101b4dccd -[FacebookConnectPlugin login:] + 973
	15  Massa                               0x0000000101ca31c6 -[CDVCommandQueue execute:] + 870
	16  Massa                               0x0000000101a1da17 -[CDVWKWebViewEngine handleCordovaMessage:] + 199
	17  Massa                               0x0000000101a1d897 -[CDVWKWebViewEngine userContentController:didReceiveScriptMessage:] + 183
	18  Massa                               0x0000000101a1f25c -[CDVWKWeakScriptMessageHandler userContentController:didReceiveScriptMessage:] + 124
	19  WebKit                              0x000000010400c042 _ZN28ScriptMessageHandlerDelegate14didPostMessageERN6WebKit12WebPageProxyERKNS0_13FrameInfoDataERN7WebCore21SerializedScriptValueE + 190
	20  WebKit                              0x0000000103fa2a65 _ZN6WebKit29WebUserContentControllerProxy14didPostMessageERN3IPC10ConnectionEyRKNS_13FrameInfoDataEyRKNS1_13DataReferenceE + 193
	21  WebKit                              0x0000000103fa4b2d _ZN3IPC13handleMessageIN8Messages29WebUserContentControllerProxy14DidPostMessageEN6WebKit29WebUserContentControllerProxyEMS5_FvRNS_10ConnectionEyRKNS4_13FrameInfoDataEyRKNS_13DataReferenceEEEEvS7_RNS_7DecoderEPT0_T1_ + 227
	22  WebKit                              0x0000000103d9b801 _ZN3IPC18MessageReceiverMap15dispatchMessageERNS_10ConnectionERNS_7DecoderE + 127
	23  WebKit                              0x0000000103f6e2da _ZN6WebKit15WebProcessProxy17didReceiveMessageERN3IPC10ConnectionERNS1_7DecoderE + 24
	24  WebKit                              0x0000000103d5e909 _ZN3IPC10Connection15dispatchMessageENSt3__110unique_ptrINS_7DecoderENS1_14default_deleteIS3_EEEE + 119
	25  WebKit                              0x0000000103d6117c _ZN3IPC10Connection18dispatchOneMessageEv + 176
	26  JavaScriptCore                      0x0000000111155299 _ZN3WTF7RunLoop11performWorkEv + 169
	27  JavaScriptCore                      0x0000000111155552 _ZN3WTF7RunLoop11performWorkEPv + 34
	28  CoreFoundation                      0x0000000106532101 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	29  CoreFoundation                      0x00000001065d1f71 __CFRunLoopDoSource0 + 81
	30  CoreFoundation                      0x0000000106516a19 __CFRunLoopDoSources0 + 185
	31  CoreFoundation                      0x0000000106515fff __CFRunLoopRun + 1279
	32  CoreFoundation                      0x0000000106515889 CFRunLoopRunSpecific + 409
	33  GraphicsServices                    0x000000010a34a9c6 GSEventRunModal + 62
	34  UIKit                               0x000000010735a5d6 UIApplicationMain + 159
	35  Massa                               0x000000010196f441 main + 65
	36  libdyld.dylib                       0x0000000109697d81 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException```

**Comment by peterpeterparker** [comment]
@fzamperin do you face the error with previous version of the plugin v1.9.1 too? is your app open source aka could you provide a link to a repo?

**Comment by fzamperin** [comment]
@peterpeterparker Thanks for your fast response, actually my app is private, but this is the code my team developed to do the login, there are 3 functions that I use to authenticate, this code works pretty well in cordova-android (6.4.0 and 7.0.0). I just got updated to v1.10.0 of the plugin right now, but even with 1.9.1 the crash persists. Although it's a native issue I use version 4.5.3 of `@ionic-native/facebook`

```
  public authenticateFacebook(): Observable<any> {
    if (this.plt.is('cordova')) {
      return Observable.create(observer => {
        //Tenta logar com o faceboo Native
        this.fb.login(['email', 'public_profile', 'user_location', 'user_birthday']).then(
          (login) => {
            console.log("Login do Facebook:", login)
            //Se logar com o facebook native, tenta logar com esses dados no servidor
            this._loginFacebook(login.authResponse).subscribe(
              logado => {
                //Se logou(a consta já existia), encerra o observer e envia a true
                if (logado) {
                  observer.next(true);
                  observer.complete();
                } else {
                  //Se não logou(então é para cadastrat), pega os dados do facebook
                  this._getFacebookData(login).subscribe(
                    facebookData => {
                      //Se conseguiu os dados do facebook, encerra e o observer e envia os dados do facebook
                      console.log("Facebook Data", facebookData);
                      facebookData.accessToken = login.authResponse.accessToken;
                      observer.next(facebookData);
                      observer.complete()
                    },
                    error => {
                      console.log("Facebook erro getData", error)
                      observer.error(error);
                    }
                  )
                }
              },
              error => {
                console.log(JSON.stringify(error));
                observer.error(error);
              }
            )
          },
          (error) => {
            //Erro ao logar com o facebook native
            console.log("facebook native error login", error)
            observer.error(error);
          }
        )
      })
    }
    else {
      return Observable.throw("É preciso estár em um device!");
    }
  }

  private _loginFacebook(authData): Observable<boolean> {
    return Observable.create(observer => {
      this.getDeviceId().subscribe(deviceId => {
        console.log(deviceId);
        this.headers = new HttpHeaders({
          'Content-Type': 'application/json'
        });
        let platform = this.plt.is('android') ? "android" : "ios"
        this.http.post(this.global.APIURL + '/auth/login/facebook', { "facebookid": authData.userID, "facebooktoken": authData.accessToken, "devicetoken": deviceId, "tipodevice": platform }, { headers: this.headers })
          .subscribe(data => {
            console.log("Login server com Facebook:", data)
            let token = (data as any).token;
            let pessoa = (data as any).pessoa;
            if (token && pessoa) {
              this.storage.set('token', token).then((val) => {
                this.storage.set('pessoa', pessoa).then((val) => {
                  this.global.PESSOA = pessoa;
                  this.global.TOKEN = token;
                  observer.next(true);
                  observer.complete();
                })
              })
            } else {
              observer.next(false);
              observer.complete();
            }
          },
          err => {
            err = err.error ? JSON.parse(err.error) : { error: 'Erro ao conectar com o servidor!', message: 'Erro desconhecido!' }
            observer.error((err as any).message);
          })
      })
    })
  }

  private _getFacebookData(authData): Observable<any> {
    return Observable.create(observer => {
      let param = ['id', 'name', 'about', 'email', 'location', 'birthday']
      this.fb.api("/" + authData.authResponse.userID + "?fields=" + param.join(), ['email', 'public_profile', 'user_location', 'user_birthday']).then(
        facebooData => {
          //Pega a imagem de perfil do facebook
          this.fb.api('me/picture?fields=url&redirect=false&width=320&height=320', []).then(
            foto => {
              if (foto.data && foto.data.url)
                facebooData.avatar = foto.data.url;
              observer.next(facebooData);
              observer.complete()
            },
            error => {
              console.log("facebook erro no get iamgem", error)
              observer.error(error);
            }
          )

        },
        error => {
          console.log("facebook erro no get user data", error)
          observer.error(error);
        }
      )
    })
  }
```

**Comment by peterpeterparker** [comment]
@fzamperin if you face the same problem using v1.9.1 then I kind of have the feeling that your issue isn't related to this issue or neither the plugin itself

since you are using ionic-native v4.5.3 I would suggest firstly to upgrade ionic-native

if still doesn't work, I would suggest you to debug what `this.fb.login` gives you as an answer

if still doesn't work, feel free to use stackoverflow or ionic forum to open a dedicated issue

**Comment by fzamperin** [comment]
@peterpeterparker I updated ionic-native to latest version and got the same issue, I'm going to debug it more carefully and if I find something I'll let you know, thanks for your help.

**Comment by darkthsideous** [comment]
@peterpeterparker  Sorry for getting back to you so late. Well it took a while for apple developer support to get back to me. They confirmed that the plugin does stall on iOS 11.3. However i tested the plugin on all earlier verions of iOS works fine. Just the latest update causing issues

**Comment by peterpeterparker** [comment]
@darkthsideous you faced the problem with v1.9.1 and/or v1.10.0 (master branch) of the plugin?

after Apple told you it wasn't ok, were you able to reproduce the problem? 

could you provide more details about the error? which call doesn't work precisely?

**Comment by darkthsideous** [comment]
@peterpeterparker seems i am still runing v1.9.1. Even after running masterbranch

`cordova plugin add cordova plugin add https://github.com/jeduan/cordova-plugin-facebook4.git --save --variable APP_ID="123456789" --variable APP_NAME="myApplication"`

Very possible i am doing wrong.

**Comment by peterpeterparker** [comment]
@darkthsideous the master branch v1.10.0 has been published to npm so you just remove/add the plugin again it should works (it worked for me, just remove it first to be sure)

    cordova plugin rm cordova-plugin-facebook4 --save
    cordova plugin add cordova-plugin-facebook4 --save --variable APP_ID="123456789" --variable APP_NAME="myApplication"

**Comment by darkthsideous** [comment]
@peterpeterparker running that command then running `cordova plugin list` still gives me v1.9.1 even when i specify `cordova-plugin-facebook4@1.10.0` Am i possibly installing wrong?

**Comment by peterpeterparker** [comment]
@darkthsideous no you are right, it's a mistake from my side sorry for that, gonna republish a version now

**Comment by darkthsideous** [comment]
@peterpeterparker no problem boss. Standing by

**Comment by peterpeterparker** [comment]
@darkthsideous it should be good now, I have just release v1.10.1 to fix the versioning number problem

sorry for the inconvenience

**Comment by darkthsideous** [comment]
@peterpeterparker v1.10.1 successfully installed going to run a test on a iOS 11 device hopefully i don't have to make too many changes to my code

**Comment by peterpeterparker** [comment]
@darkthsideous no code changes should be needed

**Comment by darkthsideous** [comment]
@peterpeterparker alright i tried to get to the bottom of this issue. Getting to the following error 
` Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process`

When i build with out the plugin .. It builds fine 

`In module 'FBSDKShareKit' imported from Numbers/Plugins/cordova-plugin-facebook4/FacebookConnectPlugin.h:14:
Numbers/Plugins/cordova-plugin-facebook4/FBSDKShareKit.framework/Headers/FBSDKAppInviteDialog.h:38:1: note: 'showFromViewController:withContent:delegate:' has been explicitly marked deprecated here
+ (instancetype)showFromViewController:(UIViewController *)viewController
^
1 warning generated.
`

**Comment by peterpeterparker** [comment]
@darkthsideous sorry I don't understand you. Are you facing a problem or not? It's a bit unclear to me

 the one stacktrace that's you point out "FBSDKAppInviteDialog.h:38:1" seems to be a warning

if you face a problem could you plz:

- Provide a link to your repo
- Document the concrete method which end up in error respectively identify and document which method of this plugin doesn't work for you
- Which version of the plugin you use
- On which OS and version you face the error

Plz provide all these information in a new issue

**Comment by peterpeterparker** [comment]
@darkthsideous 

- are you facing an error while trying the use the function "appInvite"? is that your problem?

**Comment by darkthsideous** [comment]
@peterpeterparker Thank you for getting back to me. DO you want me to open another issue. I'm using v1.10.1. It is failing to build when running
 `cordova build ios`
The project i private but this is essentially the function [here](https://ionicacademy.com/ionic-facebook-login/)

**Comment by peterpeterparker** [comment]
@darkthsideous thx for the information

in that case I think you should open a Stackoverflow or Ionic Forum thread because to me it looks more like a support questino/problem related to your app than an error related to the plugin itself

Feel free to link the URL of your thread to this isssue

**Comment by darkthsideous** [comment]
@peterpeterparker Yeah i was thinking that . Thank you much you've been a huge help!

**Comment by peterpeterparker** [comment]
@darkthsideous cool thx for your understanding, good luck and don't forget to link the URL thread here

**Comment by darkthsideous** [comment]
@peterpeterparker sure thing. May i ask what version of cordova and iOS you're using. Might help a bit

**Comment by darkthsideous** [comment]
@peterpeterparker  also your node version?
`node -v`

**Comment by peterpeterparker** [comment]
@darkthsideous 

cli packages:

    @ionic/cli-utils  : 1.19.2
    ionic (Ionic CLI) : 3.20.0

global packages:

    cordova (Cordova CLI) : 8.0.0 
    Gulp CLI              : CLI version 3.9.1 Local version 3.9.1

local packages:

    @ionic/app-scripts : 3.1.8
    Cordova Platforms  : android 7.1.0 ios 4.5.4
    Ionic Framework    : ionic-angular 3.9.2

System:

    Android SDK Tools : 26.1.1
    ios-deploy        : 1.9.2 
    ios-sim           : 5.0.13 
    Node              : v9.11.1
    npm               : 5.8.0 
    OS                : macOS High Sierra
    Xcode             : Xcode 9.3 Build version 9E145
